basque.1
.TH BASQUE 1
.SH NAME
basque \- the C Basque compiler
.SH SYNOPSIS
.B basque
[ options ] sourcefile [ options ]
.P
.B basque
[
.B \-h
|
.B \-v
]
.SH DESCRIPTION
Basque intends to be a general-purpose statically-typed procedural programming
language, similar to C, but with some improvements.
The C Basque compiler currently statically compiles for Linux x86_64 ELF, ahead
of time, and takes ASCII input.
.PP
When the C Basque compiler is invoked with an source file argument, it will
compile and assemble the source code into an executable file (there is no linker
or preprocessor).
.PP
A single dash
.B \-
can be used as the source file argument to read source code
from standard input.
.SH OPTIONS
.TP
.BR \-h ", " \-\-help
Write a list of options to standard output, then exit.
.TP
.BR \-v ", " \-\-version
Write the current version of the C Basque compiler to standard output, then exit.
.TP
.BR \-o " file"
Output compiled code into an executable file at the specified path. To write to
standard output, use a dash
.B \-
as the
.B file
argument. If this argument is not provided, and the input file name is
.BR filename.ba ","
the output file will be
.BR filename "."
Otherwise, the output file will be called
.BR out "."
.TP
.BR \-r ", " \-\-run
Run the compiled code after compilation. This will still generate an output file.
.TP
.BR \-s ", " \-\-silence\-warnings
Silence all compiler warnings.
.TP
.BR \-W ", " \-\-warnings-as-errors
Treat warnings as errors (terminate compilation on warnings).
.TP
.BR \-\-disable\-assertions
Disable compiling code for
.B assert
statements.
.TP
.BR \-\-include\-path " path"
Set paths for searching for included files. When an include statement is
processed, the C Basque compiler will first search in the same directory as the
file where the statement was written, then will search in include paths passed
to the compiler, in order from the start to the end of the string passed. Paths 
are separated by colons.
.TP
.BR \-\-page\-size " size"
Set the size of memory pages in bytes. Setting the page size to 0 will leave it as default.
.SH COPYRIGHT
Copyright (C) 2021-2022 nikolashn.
.PP
The C Basque compiler is distributed under the terms of the GNU General Public License version 3, as published by the Free Software Foundation.


basque.c
// See LICENSE for copyright/license information

#include "lexer.h"
#include "bltin/bltin.h"
#include "parser/parse.h"
#include "elf64.h"

char usageStr[] = 
	"Usage: basque [options] file\n"
	"Use \"-\" for file to read from stdin.\n"
	"Options:\n"
	"  -h,--help               Displays this text.\n"
	"  -v,--version            Display the current version of the C Basque compiler.\n"
	"  -o <FILE>               Output compiled code into <FILE>. Use - for stdout.\n"
	"  -r,--run                Run the compiled code after compilation.\n"
	"  -s,--silence-warnings   Silence warnings.\n"
	"  -W,--warnings-as-errors Treat warnings as errors.\n"
	"  --disable-assertions    Disable compiling code for \"assert\" statements.\n"
	"  --include-path <PATH>   Set paths for searching for included files.\n"
	"  --page-size <SIZE>      Size in bytes of memory pages.\n"
;

int main(int argc, char* argv[]) {
	// ----- Handle args ------
	
	if (argc < 2) {
		fprintf(stderr, "%s", usageStr);
		return 0;
	}

	FILE* srcFile = 0;
	char* srcFileName = 0;
	char* outFileName = 0;

	bool isRunCode = 0;

	for (u64 i = 1; i < argc; i++) {
		bool isHandledCLO = 0;
		bool isCLOHelp = 0;
		bool isCLOVersion = 0;
		
		u64 len = strlen(argv[i]);
		
		if (!strcmp(argv[i], "-o")) {
			if (++i == argc) {
				fprintf(stderr, "Error: no file name provided after -o\n");
				return 1;
			}
			outFileName = argv[i];
			goto BA_LBL_MAIN_ARGSLOOPEND;
		}
		else if (argv[i][0] == '-') {
			if (len > 1 && argv[i][1] != '-') {
				isHandledCLO = 1;
				for (u64 j = 1; j < len; j++) {
					switch (argv[i][j]) {
						case 'h':
							isCLOHelp = 1;
							break;
						case 'v':
							isCLOVersion = 1;
							break;
						case 'r':
							isRunCode = 1;
							break;
						case 's':
							ba_SetSilenceWarns();
							break;
						case 'W':
							ba_SetWarnsAsErrs();
							break;
						default:
							fprintf(stderr, "Error: Command line option %s "
								"not found\n", argv[i]);
							return 1;
					}
				}
			}
		}
		
		if (isCLOHelp || !strcmp(argv[i], "--help")) {
			fprintf(stderr, "%s", usageStr);
			return 0;
		}
		else if (isCLOVersion || !strcmp(argv[i], "--version")) {
			fprintf(stderr, "basque " BA_VERSION "\n");
			return 0;
		}
		else if (!strcmp(argv[i], "--run")) {
			isRunCode = 1;
		}
		else if (isHandledCLO) {
			goto BA_LBL_MAIN_ARGSLOOPEND;
		}
		else if (!strcmp(argv[i], "--silence-warnings")) {
			ba_SetSilenceWarns();
		}
		else if (!strcmp(argv[i], "--warnings-as-errors")) {
			ba_SetWarnsAsErrs();
		}
		else if (!strcmp(argv[i], "--page-size")) {
			if (++i == argc) {
				fprintf(stderr, "Error: no page size provided after --page-size\n");
				return 1;
			}
			ba_SetPageSize(atoll(argv[i]));
			goto BA_LBL_MAIN_ARGSLOOPEND;
		}
		else if (!strcmp(argv[i], "--disable-assertions")) {
			ba_SetAssertions(0);
			goto BA_LBL_MAIN_ARGSLOOPEND;
		}
		else if (!strcmp(argv[i], "--include-path")) {
			if (++i == argc) {
				fprintf(stderr, "Error: no path provided after --include-path\n");
				return 1;
			}
			ba_SetIncludePath(argv[i], /* isOverrideDefaults = */ 0);
			goto BA_LBL_MAIN_ARGSLOOPEND;
		}
		else if (len > 1 && argv[i][0] == '-' && argv[i][1] == '-') {
			fprintf(stderr, "Error: Command line option %s not found\n", argv[i]);
			return 1;
		}
		else if (!srcFile) {
			if (!strcmp(argv[i], "-")) {
				srcFileName = ba_MAlloc(2);
				srcFileName[0] = '.';
				srcFileName[1] = 0;
				srcFile = stdin;
			}
			else {
				srcFileName = ba_MAlloc(strlen(argv[i])+1);
				strcpy(srcFileName, argv[i]);
				srcFile = fopen(srcFileName, "r");
			}
			if (!srcFile) {
				fprintf(stderr, "Error: Cannot find %s: no such file\n", argv[i]);
				return 1;
			}
			if (!outFileName) {
				if ((len > 3) && !strcmp(argv[i]+len-3, ".ba")) {
					outFileName = argv[i];
					outFileName[len-3] = 0;
				}
				else {
					outFileName = "out";
				}
			}
		}
		else {
			fprintf(stderr, "Error: only one input file may be specified\n");
			return 1;
		}
		BA_LBL_MAIN_ARGSLOOPEND:;
	}
	if (ba_IsWarnsAsErrs() && ba_IsSilenceWarns()) {
		fprintf(stderr, "Error: cannot both silence warnings and have warnings "
			"as errors\n");
		return 1;
	}
	if (!srcFile) {
		fprintf(stderr, "%s", usageStr);
		return 0;
	}

	// ----- Begin parsing -----
	
	struct ba_Ctr* ctr = ba_NewCtr();
	ctr->currPath = srcFileName;
	ctr->dir = ba_MAlloc(strlen(srcFileName)+1);
	strcpy(ctr->dir, srcFileName);
	ctr->dir = dirname(ctr->dir);

	if (!ba_Tokenize(ctr, srcFile)) {
		return 1;
	}
	fclose(srcFile);
	
	// DEBUG
	/*
	struct ba_Lexeme* lex = ctr->startLex;
	while (lex) {
		fprintf(stderr, "%llx %s\n", lex->type, lex->val);
		lex = lex->next;
	}
	*/

	if (!ba_Parse(ctr)) {
		return 1;
	}

	// ----- Optimization -----
	// TODO: optimization

	// ----- Binary generation -----
	
	if (!ba_WriteBinary(outFileName, ctr)) {
		return 1;
	}

	if (isRunCode) {
		char* runFileName = outFileName;
		// Relative path
		if (outFileName[0] != '/') {
			// Add "./" to the start
			runFileName = ba_MAlloc(strlen(outFileName)+2);
			runFileName[0] = '.';
			runFileName[1] = '/';
			runFileName[2] = 0;
			strcat(runFileName, outFileName);
		}
		// File cannot be read or executed
		if (access(outFileName, R_OK|X_OK)) {
			fprintf(stderr, "Error: cannot read or cannot execute "
				"written binary file\n");
			return 1;
		}
		char* args[] = { runFileName, NULL };
		execvp(runFileName, args);
	}

	free(ba_GetIncludePath());
	
	return 0;
}

